src/06_ffi.md

```diff
@@ -0,0 +1,395 @@
+
+Note that `catch_unwind` will only catch unwinding panics, not those that abort
+the process.
+
+```rust ignore
+use std::panic::catch_unwind;
+use rand;
+
+pub fn may_panic() {
+    if rand::random() {
+        panic!("panic happens");
+    }
+}
+
+#[no_mangle]
+pub extern fn no_panic() -> i32 {
+    let result = catch_unwind(||may_panic());
+    match result {
+        Ok(_) => 0,
+        Err(_) => -1,
+    }
+}
+```
+
+### `no_std`
+
+In the case of `#![no_std]` program, a panic handler (`#[panic_handler]`) must
+be defined to ensure security. The panic handler should be written with great
+care in order to ensure both the safety and security of the program.
+
+Another approach is to simply ensure that there is no use of `panic!` with the
+[`panic-never`] crate. Like [`no-panic`], [`panic-never`] relies on a linking
+trick: the linker fails if a non-trivially-dead branch leads to `panic!`.
+
+[`panic-never`]: https://crate.io/crates/panic-never
+[`no-panic`]: https://github.com/dtolnay/no-panic
+
+## Binding a foreign library in Rust
+
+> ### Recommendation {{#check FFI-SAFEWRAPPING | Provide safe wrapping to foreign library}}
+>
+> Interfacing a library written in another language in Rust should be done in
+> two parts:
+>
+> - a low-level, possibly *hidden*, module that closely translates the original
+>   C API into `extern` blocks,
+> - a safe wrapping module that ensures memory safety and security invariants at
+>   the Rust level.
+>
+> If the low-level API is exposed to the world, it should be done in a dedicated
+> crate with a name of the form `*-sys`.
+
+The crate [rust-bindgen] may be used to automatically generate the low-level
+part of the binding from C header files.
+
+<mark>TODO</mark> example
+
+## Binding a Rust library in another language
+
+> ### Recommendation {{#check FFI-CAPI | Expose dedicated C-compatible API only}}
+>
+> In a secure Rust development, exposing a Rust library to a foreign language
+> should only be done through a **dedicated C-compatible API**.
+
+The crate [cbindgen] may be used to automatically generate C or C++ bindings to
+the Rust C-compatible API of a Rust library.
+
+### Minimal example of a C-exported Rust library
+
+`src/lib.rs`:
+
+```rust
+/// Opaque counter
+pub struct Counter(u32);
+
+impl Counter {
+    /// Create a counter (initially at 0)
+    fn new() -> Self {
+        Self(0)
+    }
+    /// Get the current value of the counter
+    fn get(&self) -> u32 {
+        self.0
+    }
+    /// Increment the value of the counter if there's no overflow
+    fn incr(&mut self) -> bool {
+        if let Some(n) = self.0.checked_add(1) {
+            self.0 = n;
+            true
+        } else {
+            false
+        }
+    }
+}
+
+// C-compatible API
+
+#[no_mangle]
+pub extern fn counter_create() -> *mut Counter {
+    Box::into_raw(Box::new(Counter::new()))
+}
+
+#[no_mangle]
+pub extern fn counter_incr(counter: *mut Counter) -> std::os::raw::c_int {
```

Comment:
These should be marked `unsafe`; let's not forget that `#[no_mangle] pub extern "C"` functions can still be called from within Rust! (I guess a warning about forgetting this "detail" could be added somewhere in the guide).

Reply:
I see your point. But at the same time, I am not sure about that.

First, using `unsafe` at the function level means also switching off safety very early. There is no `safe` block that may counteract this annotation.

Second, in the wild, it's unclear:

* Examples in the Book, the Nomicon, etc. do not do this
* Some C APIs do: encoding_c (used in Firefox)
* Some do not: rvs-c-api, 
* Some do it sometimes (!): wasmer-runtime-c-api, webrender binding in Firefox

In fact in most case `unsafe` is more used as a way to mark the whole function body unsafe than to instruct the caller.

Third, IMO the goal here is to have extern functions as robust as possible. If they cannot work in a safe context when would they work?

I think the more useful rec is to write a dedicated C API that cannot be easily confused with the Rust standard API.

Reply:
The function **has** to be marked `unsafe`, since it has a narrow contract. See https://github.com/rust-lang/unsafe-code-guidelines/pull/171

On the other hand, I agree that no longer requiring `unsafe` blocks for the most delicate parts is not great;

> There is no safe block that may counteract this annotation.

Luckily, there is a way to do it:

```rust
/// # Safety
///
///   - `counter` must be `NULL` or point to a read-writeable `Counter` instance.
///
///   - must not be called from multiple threads in parallel (not thread-safe).
#[no_mangle]
pub
unsafe // narrow contract!
extern "C"
fn counter_incr (counter: *mut Counter) -> ::std::os::raw::c_int
{
    /// The inner function is private so it does not need to be marked `unsafe`
    #[inline]
    fn counter_incr (counter: *mut Counter) -> ::std::os::raw::c_int
    {
        if let Some(counter) = unsafe { counter.as_mut() } {
            if counter.incr() {
                0
            } else {
                -1
            }
        } else {
            -2
        }
    }
    counter_incr(counter)
}
```

I may publish a `proc_macro_attribute` `#[require_unsafe_blocks_in_body]` to automate using this pattern.

Reply:
I thought about using a helper function but it's quite cumbersome.
#[require_unsafe_blocks_in_body]` is a great idea!

The consistency with the unsafe guide is a plus too

Reply:
[`#[require_unsafe_in_body]`](https://crates.io/crates/require_unsafe_in_body) has been released :)